pkg/kratos/handlers.go

```diff
@@ -137,42 +165,118 @@ func (a *API) handleUpdateFlow(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	loginFlow, _, err := a.service.GetLoginFlow(context.Background(), flowId, r.Cookies())
+	loginFlow, cookies, err := a.service.GetLoginFlow(r.Context(), flowId, r.Cookies())
 	if err != nil {
 		a.logger.Errorf("Error when getting login flow: %v\n", err)
 		http.Error(w, "Failed to get login flow", http.StatusInternalServerError)
 		return
 	}
 
-	allowed, err := a.service.CheckAllowedProvider(context.Background(), loginFlow, body)
+	allowed, err := a.service.CheckAllowedProvider(r.Context(), loginFlow, body)
 	if err != nil {
 		a.logger.Errorf("Error when authorizing provider: %v\n", err)
 		http.Error(w, "Server error", http.StatusInternalServerError)
 		return
 	}
+
 	if !allowed {
 		http.Error(w, "Provider not allowed", http.StatusForbidden)
 		return
 	}
 
-	flow, cookies, err := a.service.UpdateLoginFlow(context.Background(), flowId, *body, r.Cookies())
+	flow, cookies, err := a.service.UpdateLoginFlow(r.Context(), flowId, *body, r.Cookies())
 	if err != nil {
 		a.logger.Errorf("Error when updating login flow: %v\n", err)
 		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
 
-	resp, err := json.Marshal(flow)
+	shouldEnforceMfa, err := a.shouldEnforceMFA(r.Context(), cookies)
 	if err != nil {
-		a.logger.Errorf("Error when marshalling Json: %v\n", err)
-		http.Error(w, "Failed to parse login flow", http.StatusInternalServerError)
+		err = fmt.Errorf("enforce check error: %v", err)
+		a.logger.Error(err.Error())
+		http.Error(w, err.Error(), http.StatusInternalServerError)
 		return
 	}
+
 	setCookies(w, cookies)
-	// Kratos returns us a '422' response but we tranform it to a '200',
-	// because this is the expected behavior for us.
+
+	if shouldEnforceMfa {
+		a.mfaSettingsRedirect(w)
+		return
+	}
+
 	w.WriteHeader(http.StatusOK)
-	w.Write(resp)
+	_ = json.NewEncoder(w).Encode(flow)
+}
+
+func (a *API) shouldEnforceMFA(ctx context.Context, cookies []*http.Cookie) (bool, error) {
+	if !a.mfaEnabled {
+		return false, nil
+	}
+
+	session, _, err := a.service.CheckSession(ctx, cookies)
+	if err != nil {
+		if a.is40xError(err) {
+			a.logger.Debugf("check session failed, err: %v", err)
```

Comment:
if I understand correctly, this is not a real error because we let kratos redirect to the 2fa screen as specified in config instead of enforcing it?

Reply:
Yes, because we can't access the `whoami` endpoint so we get a 401 or 403, in that case we don't have to enforce mfa, because we have to do it after the actual login happen and we get the session to check the necessary stuff